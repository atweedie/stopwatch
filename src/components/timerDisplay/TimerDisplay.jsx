import { useContext, useEffect } from "react";
import { TimerStatus } from "../../modules/timerStatusContext";

const TimerDisplay = () => {
    const {timerState, setTimerState} = useContext(TimerStatus)

    const date = new Date(timerState.elapsedTime);

    useEffect(() => {
        if (timerState.isRunning) {
            setTimeout(() => {
                setTimerState({
                    elapsedTime: timerState.elapsedTime + 10,
                    isRunning: timerState.isRunning
                })
            }, 10);
        }
    })

    return (
        <h1>
            {date.toISOString().substring(12,22)}
        </h1>
    )
}
    

export default TimerDisplay;