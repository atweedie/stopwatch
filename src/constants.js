export const TIMER_TICK_DURATION_IN_MS = 10;