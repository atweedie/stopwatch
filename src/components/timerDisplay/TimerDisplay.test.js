import React from 'react'
import { render } from '@testing-library/react';
import TimerDisplay from './TimerDisplay';
import {TIMER_TICK_DURATION_IN_MS} from "../../constants";

const stubSetElapsedTime = jest.fn();
const stubSetIsRunning = jest.fn();
const stubUseContext = jest.fn();

const defaultContext = {
    elapsedTime: 0,
    isRunning: false,
    setElapsedTime: stubSetElapsedTime,
    setIsRunning: stubSetIsRunning
}

const renderWithContext = (context = defaultContext) => {
    React.useContext = stubUseContext.mockReturnValue(context)
    return render(<TimerDisplay/>)
};

beforeEach(() => {
    jest.useFakeTimers();
})

afterEach(() => {
    jest.useRealTimers();
});

test('should render as expected', () => {
const component = renderWithContext();
    
    expect(component.container).toMatchSnapshot();
});

test('should not call setElapsedTime if isRunning is false', () => {
    const givenContext = {
        ...defaultContext,
        isRunning: false,
    };

    renderWithContext(givenContext);
    
    jest.runAllTimers();

    expect(stubSetElapsedTime).not.toHaveBeenCalled();
})

test('should call setElapsedTime with expected parameters if isRunning is true', () => {
    const givenContext = {
        ...defaultContext,
        isRunning: true,
    };

    renderWithContext(givenContext);
    
    jest.runAllTimers();

    expect(stubSetElapsedTime).toHaveBeenCalledWith(TIMER_TICK_DURATION_IN_MS);
})

