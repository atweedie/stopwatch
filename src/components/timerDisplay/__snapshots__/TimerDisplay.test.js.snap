// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as expected 1`] = `
<div>
  <h1
    class="timer"
    id="stopwatch-elapsed"
  >
    0:00:00.00
  </h1>
</div>
`;
